programs/candy-guard/program/src/guards/asset_payment_multi.rs

```diff
@@ -0,0 +1,106 @@
+use mpl_core::instructions::TransferV1CpiBuilder;
+
+use super::*;
+
+use crate::{state::GuardType, utils::assert_keys_equal};
+
+/// Guard that requires a specific number of Assets from a specific collection to be used as payment.
+///
+/// List of accounts required:
+///
+///   0. `[writeable]` Address of the collection.
+///   1. `[writeable]` Address of the destination.
+///   x. `[writeable]` Address of the Asset(s).
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
+pub struct AssetPaymentMulti {
+    pub required_collection: Pubkey,
+    pub destination: Pubkey,
+    pub num: u8,
+}
+
+impl Guard for AssetPaymentMulti {
+    fn size() -> usize {
+        32 // required_collection
+        + 32 // destination
+        + 1 // num of assets to burn
+    }
+
+    fn mask() -> u64 {
+        GuardType::as_mask(GuardType::AssetPaymentMulti)
+    }
+}
+
+impl Condition for AssetPaymentMulti {
+    fn validate<'info>(
+        &self,
+        ctx: &mut EvaluationContext,
+        _guard_set: &GuardSet,
+        _mint_args: &[u8],
+    ) -> Result<()> {
+        let index = ctx.account_cursor;
+        let collection_account = try_get_account_info(ctx.accounts.remaining, index)?;
+        let destination_account = try_get_account_info(ctx.accounts.remaining, index + 1)?;
+        ctx.account_cursor += 2;
+
+        let mut i: usize = 0;
+        let mut asset_account;
+        let mut asset;
+
+        assert_keys_equal(&destination_account.key(), &self.destination)?;
+
+        while i < usize::from(self.num) {
```

Comment:
This is fine just some rust tips.

This is slightly hardcore for a number cast, I think normally you just see `self.num as usize`.

Also could use a for loop `for i in 0..self.num as usize`